crates/cranelift/src/gc/enabled.rs

```diff
@@ -79,6 +389,15 @@ impl FuncEnvironment<'_> {
 
     /// Get the raw pointer of `gc_ref[offset]` bounds checked for an access of
     /// `size` bytes.
+    ///
+    /// The given `gc_ref` must be a non-null, non-i31 GC reference.
+    ///
+    /// Returns the raw pointer to `gc_ref[offset]` -- not a raw pointer to the
+    /// GC object itself (unless `offset == 0`). This raw pointer may be used to
+    /// read or write up to `size` bytes. Do NOT attempt accesses larger than
+    /// `size` bytes; that may lead to unchecked out-of-bounds accesses.
+    ///
+    /// This method is collector-agnostic.
     fn prepare_gc_ref_access(
```

Comment:
This method may want to be refactored to return `(ir::Value, u32)` to enable optionally embedding the offset in the load/store instruction. For now it'd always return 0 but maybe a future refactoring to have to optimize this a bit in the future.

Reply:
Yeah, I was also thinking it may want to take a `object_size: Option<u32>` argument as well to make bounds checks of repeated accesses dedupe-able by the mid-end. (Has to be an `Option` because we don't statically know array lengths).

Reply:
But: simple, naive things first and we can clean up the code we generate in follow ups.